apps/src/blockly/addons/cdoFieldPicker.js

```diff
@@ -0,0 +1,129 @@
+ * @param buttonIcon If the field displyas a button, this is the string that is displayed on the button.
+ */
+export default class CdoFieldPicker extends GoogleBlockly.Field {
+  constructor(value, onChange, onDisplay, buttonIcon) {
+    super(value);
+    this.onChange = onChange;
+    this.onDisplay = onDisplay;
+    this.buttonIcon = buttonIcon;
+
+    this.SERIALIZABLE = true;
+  }
+
+  static fromJson(options) {
+    return new CdoFieldPicker(
+      options.value,
+      options.onChange,
+      options.onDisplay,
+      options.buttonIcon
+    );
+  }
+
+  /**
+   * Validate the changes to a field's value before they are set.
+   * @override
+   */
+  doClassValidation_(newValue) {
+    if (typeof newValue !== 'string') {
+      return null;
+    }
+    // Handle 'play sound' block with default param from CDO blockly.
+    // TODO: Remove when sprite lab is migrated to Google blockly.
+    if (newValue === 'Choose') {
+      return DEFAULT_SOUND;
+    }
+    return newValue;
+  }
+
+  /**
+   * Create the block UI for this field.
+   * @override
+   */
+  initView() {
+    super.initView();
+    if (this.buttonIcon) {
+      this.buttonElement_ = Blockly.utils.dom.createSvgElement(
+        'rect',
+        {
+          rx: BUTTON_CORNER_RADIUS,
+          ry: BUTTON_CORNER_RADIUS,
+          x: 1,
+          y: 1,
+          height: BUTTON_INNER_HEIGHT,
+          width: BUTTON_INNER_HEIGHT,
+        },
+        this.fieldGroup_
+      );
+      this.buttonElement_.style.fillOpacity = 1;
+      this.buttonElement_.style.fill =
+        this.getSourceBlock().style.colourSecondary;
+
+      this.textElement_.style.fontSize = '11pt';
+      this.textElement_.style.fill = this.getSourceBlock().style.colourPrimary;
+      this.textElement_.textContent = '';
+      this.textElement_.appendChild(this.buttonIcon);
+
+      this.fieldGroup_.insertBefore(this.buttonElement_, this.textElement_);
+    }
+  }
+
+  /**
+   *  Get the text from this field to display on the block. May differ from
+   * `getText` with call to `this.onDisplay` which can change format of text.
+   * @override
+   */
+  getDisplayText_() {
+    let text = this.getText();
+    if (!text) {
+      return GoogleBlockly.Field.NBSP;
+    }
+    // The onDisplay function customizes the text for display.
+    return this.onDisplay(text);
+  }
+
+  /**
+   * Used to update the value of a field.
+   * @override
+   */
+  doValueUpdate_(newValue) {
+    if (this.value_ !== newValue) {
+      this.value_ = newValue;
+      if (this.buttonIcon) {
+        // For location picker, update coordinates of block when dragged out from toolbox.
+        this.onDisplay(newValue);
+      }
+      this.isDirty_ = true; // Block needs to be re-rendered.
+    }
+  }
+
+  /**
+   * Create an editor for the field.
+   * @override
+   */
+  showEditor_() {
+    this.onChange();
+  }
+
+  /**
+   * Contrast background for button with source block
+   * @override
+   */
+  applyColour() {
```

Comment:
Only used for location block?

Reply:
Used for when there is a `buttonIcon` passed in the constructor.